inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/annotation/DynamicAnnotatorWorkflowActionBarItemGroup.java

```diff
@@ -89,48 +102,115 @@ public String getStateClass()
 
-            AnnotationDocument annotationDocument = documentService
-                .getAnnotationDocument(document, user);
+            AnnotationDocument annotationDocument = documentService.getAnnotationDocument(document,
+                    user);
 
             documentService.transitionAnnotationDocumentState(annotationDocument,
-                ANNOTATION_IN_PROGRESS_TO_ANNOTATION_FINISHED);
-
-            //Go through all documents in a random order and check if there is a Annotation document
-            //with the state NEW
-            String query =  "FROM SourceDocument " +
-                            "WHERE project = :project " +
-                            "ORDER BY rand()";
-            for (SourceDocument doc: entityManager.createQuery(query,SourceDocument.class)
-                .setParameter("project", project).getResultList()) {
-                //Check if it even exists or is state NEW
-                if (documentService.listAnnotatableDocuments(project,user).get(doc) == null ||
-                    documentService.listAnnotatableDocuments(project,user).get(doc).
-                        getState().equals(NEW)) {
-                    getAnnotationPage().getModelObject().setDocument(doc, documentService.
-                        listSourceDocuments(project));
-                    //This document had the state NEW, load it
+                    ANNOTATION_IN_PROGRESS_TO_ANNOTATION_FINISHED);
+
+            // FIXME: Instead of querying for ALL documents, it would be faster to have a query
+            // method to get annotation documents in a particular state...
+            List<AnnotationDocument> annotationDocumentList = documentService
+                    .listAnnotationDocuments(project, user);
+            for (AnnotationDocument anno : annotationDocumentList) {
+                // There was one in progress, load it
+                if (IN_PROGRESS.equals(anno.getState())) {
+                    getAnnotationPage().getModelObject().setDocument(anno.getDocument(),
+                            documentService.listSourceDocuments(project));
                     getAnnotationPage().actionLoadDocument(_target);
-                    _target.add(page);
                     return;
                 }
             }
 
-            //Check if no new document has been selected, return to homepage
-            if (getAnnotationPage().getModelObject().getDocument().equals(document)) {
-                getAnnotationPage().setResponsePage(getAnnotationPage().getApplication().
-                    getHomePage());
-                getSession().info("There are no more documents to annotate available for you. Please contact your project supervisor.");
+            WorkloadManager currentWorkload = workloadManagementService
+                    .getOrCreateWorkloadManagerConfiguration(project);
+
+            WorkloadTraits traits = dynamicWorkloadExtension.readTraits(currentWorkload);
```

Comment:
Instead of having a generic `WorkflowTraits` for all workflow types, there should be a `DynamicWorkflowTraits` and a `MatrixWorkflowTraits`, each with the settings required by **the particular workflow strategy**.